### PR TITLE
feat: integration development skill

### PR DESCRIPTION
**Pull request by protoss70**
# What

A new integration development skill for companies that want to build their Apify integrations. Has sections coversing Workflow automation platforms, agent harness plugins, SDK integration and AI framework integrations.

The root skill.md includes Apify info and commands the agent to fetch https://apify.com/agents.md for up to date Apify information

# How I created the skill

I created design analysis md files on:
- n8n, Zapier for workflow automation
- openclaw and hermes plugins for harnesses
- langchain for AI framework integrations
- Also replaced the old sdk integration skill with this master skill

Then I merged the information into 1 skill file

Please let me know if any info is missing

@lukas-bekr could you please let me know if it is okay for me to merge the skills and replace the old sdk integration skill?

closes https://github.com/apify/integrations-team/issues/79

### CODE REVIEW

File: skills/apify-integration-development/references/sdk-integration.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+POST https://api.apify.com/v2/actors/{username}~{actor-name}/run-sync-get-dataset-items
+```
+
+Longer work must use the asynchronous flow: POST /runs -> poll GET /actor-runs/{id} -> GET /datasets/{id}/items.
+
+REST reference: `https://docs.apify.com/api/v2`. OpenAPI spec: `https://apify.com/openapi.json`.
+
+## 7. Best practices
+
+- **Set timeouts:** pass `timeoutSecs` in the Actor input or `waitSecs` on `.call()` to avoid indefinite waits.
```
Comment (drobnikj) [inline skills/apify-integration-development/references/sdk-integration.md]: Important: `timeoutSecs` is a run option, not Actor input — putting it in the input contradicts the options-not-input rule in `SKILL.md`, and an Actor whose schema rejects unknown fields will fail on it. Point at the client's call options / the `timeout` query parameter instead.

File: skills/apify-integration-development/references/sdk-integration.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+### Asynchronous execution (start and poll)
+
+```typescript
+const run = await client.actor('apify/web-scraper').start({
+    startUrls: [{ url: 'https://example.com' }],
+});
+
+// Poll for completion
+const finishedRun = await client.run(run.id).waitForFinish();
```
Comment (drobnikj) [inline skills/apify-integration-development/references/sdk-integration.md]: Note: `waitForFinish()` with no `waitSecs` waits indefinitely, so the canonical async example breaks the bounded-polling rule that `SKILL.md` and line 212 of this same file both insist on. Give the example an explicit bound.

File: skills/apify-integration-development/references/sdk-integration.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    const run = await client.actor('apify/web-scraper').call(input);
+
+    if (run.status !== 'SUCCEEDED') {
+        const log = await client.log(run.id).get();
+        throw new Error(`Actor failed with status ${run.status}: ${log}`);
+    }
+
+    const { items } = await client.dataset(run.defaultDatasetId).listItems();
+} catch (error) {
+    if (error.message?.includes('not found')) {
```
Comment (drobnikj) [inline skills/apify-integration-development/references/sdk-integration.md]: Suggestion: matching on a message substring is brittle — `apify-client` errors carry `statusCode` and `type` (`record-not-found`), which the very next branch already uses.

File: skills/apify-integration-development/references/workflow-automation.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+Map the host's file-handling primitives (dehydration / stashing / signed URLs) to Apify's binary KV records. For "Get KV record":
+- JSON/text under a size threshold (e.g. 20 MB) -> parsed and returned as fields the user can map directly.
+- Binary or larger records -> a dehydrated pointer the host lazily fetches when a downstream step needs it.
+- Records above a hard limit (e.g. 120 MB, with margin below the host's 150 MB ceiling) -> rejected outright with an actionable message.
+
+## 10. Run-finished trigger UX
+
+Use the host's hook type backed by Apify webhooks:
+- **Subscribe**: create an Apify webhook scoped to an `actorId` or `actorTaskId`, with `eventTypes` from the user-selected terminal statuses. The webhook's `requestUrl` is the host's target URL. Make registration **idempotent** - derive a key from `resource:id:sortedEvents` so re-activating a workflow does not create duplicate webhooks. Persist the created webhook ID so deactivation can delete it.
```
Comment (drobnikj) [inline skills/apify-integration-development/references/workflow-automation.md]: Note: a key of `resource:id:sortedEvents` omits `requestUrl`, and Apify returns the *existing* webhook for a repeated `idempotencyKey` instead of creating one — so two workflows watching the same Actor and statuses would collide and the second would never fire. The key has to cover every field that distinguishes one registration from another.

File: skills/apify-integration-development/references/ai-framework-package.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+A single predictable envelope lets agents parse results with one code path. The `run` metadata gives the agent enough to chain calls - run an Actor with one tool, then fetch the dataset with another using the returned `dataset_id`. End every tool description with "Use only the data returned; do not hallucinate missing fields."
+
+## 5. Safety clamps - defense against LLM-requested extremes
+
+An LLM invoking a tool can request absurd values: 10,000 results, 32 GB of memory, a 1-hour timeout. Clamp every request to **developer-controlled ceilings**:
+
+| Clamp | Default ceiling |
+|---|---|
+| `timeout_secs` | 600 s |
+| `memory_mbytes` | 32,768 MB (snapped to nearest valid power-of-2) |
```
Comment (drobnikj) [inline skills/apify-integration-development/references/ai-framework-package.md]: Note: 32,768 MB is Apify's maximum run memory, so this ceiling clamps nothing while CU cost scales linearly with memory. A clamp meant to stop LLM-requested extremes should default well below the platform max.

File: skills/apify-integration-development/references/ai-harness-plugin.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+## 11. Dependency injection for tests
+
+The tool factory should accept an optional injected `client`. When omitted, construct a real client from the resolved key; when provided (in tests), bypass it entirely. This lets the test suite exercise every action and edge case - store search, schema fetch, run start, collect success/pending, unknown action, missing key - with no network access, using a hand-rolled mock shaped to the SDK's method-chain surface. No mocking framework needed.
+
+## 12. Known gaps to design for
+
+1. **Bill caps in the prompt, not the tool.** If `start` forwards only the Actor's `input` with no options channel, there is no `maxTotalChargeUsd` / `maxItems` cap plumbed through. The cost rule is enforced in the prompt via the README/pricing info `discover` returns. This is a known gap vs the MCP and REST paths - worth closing if the harness exposes an options argument.
```
Comment (drobnikj) [inline skills/apify-integration-development/references/ai-harness-plugin.md]: Note: `SKILL.md` makes a cost control mandatory on every run-start path, but this documents shipping without one and the checklist below has no cost-cap line. Either state the exception in the cross-cutting rule or add the requirement to the checklist — as written an agent can tick every box and still ship uncapped runs.

**Comment by drobnikj** [comment]
> Solves: [apify/integrations-team#79](https://github.com/apify/integrations-team/issues/79)

@protoss70, please use the correct keyword for connection PR otherwise it will not connect, see https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Review by drobnikj — COMMENTED** [review]
Looks good! Just a few notes.

File: skills/apify-integration-development/references/workflow-automation.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+If the host is headless-only, fall back to an API-token credential with the same verify call on a "Verify" button.
+
+## 13. Convenience operations: "Scrape single URL"
+
+Beyond generic "run Actor", ship a curated **Scrape single URL** action: a 2-field form (`url`, `outputFormat`) wrapping a content scraper with sensible defaults (`maxCrawlDepth: 0`, `maxResults: 1`). Validate the URL first (`new URL()` + protocol + hostname check) with an actionable error - bad input must never start a paid run. Return a **lean, single-object** output: `{ ...pageMetadata, [outputFormat]: content }` - strip all content variants and re-add only the chosen one. This is ideal for LLM flows (a single object beats a dataset array) and lowers the barrier for non-power users. Point power users to the underlying Actor for advanced options.
```
Comment (JanHranicky) [inline skills/apify-integration-development/references/workflow-automation.md:146]: We use WCC for scrape single url triggers, but we want to migrate towards unblocking teams Web fetch actor - https://apify.com/apify/web-fetch

I would make the skill follow this new convention. Call the Action `Web fetch` and use the correct Actor

File: skills/apify-integration-development/references/ai-harness-plugin.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+If the harness's `register()` is synchronous and the loader does not `await` it (a common gotcha), keep registration fully synchronous - build the tool (construct a client + schema, no I/O) and register inline. Any network call happens later inside a tool `execute` or CLI action, where async is expected.
+
+## 8. SDK handling and attribution
+
+Use the official `apify-client` SDK (JS or Python), not raw HTTP. Construct the client once, memoized, and rebuilt only when the token changes. Stamp the attribution header on every request: `x-apify-integration-platform: <your-harness>`. When the request is driven by the AI tool (always, in this shape), also send `x-apify-integration-ai-tool: true`. This is the single most important line for Apify's side of the relationship.
```
Comment (JanHranicky) [inline skills/apify-integration-development/references/ai-harness-plugin.md]: As per todays discussion, it might be good to add some kind of header that will enable us to distunguish a skill generated integration from a custom one. 

It is a flaky recognition system as the user might still remove the header, but better than nothing.

**Review by JanHranicky — APPROVED** [review]
LGTM, nothing blocking. Pre-approving :)

File: skills/apify-integration-development/references/workflow-automation.md (hunk 7 above)
Comment (protoss70) [reply skills/apify-integration-development/references/workflow-automation.md:146]: Created a follow up ticket for this since I am not quite sure on how the migration will look like now. Once I solve the Zapier migration to new fetch standby Actor I will update this

https://github.com/apify/integrations-team/issues/144

File: skills/apify-integration-development/references/ai-harness-plugin.md (hunk 8 above)
Comment (protoss70) [reply skills/apify-integration-development/references/ai-harness-plugin.md]: Using `x-apify-integration-origin: apify-integration-development-skill`
